ptrace/func_call.py

```diff
@@ -62,6 +62,11 @@ def __getitem__(self, key):
 
     def format(self):
         arguments = [arg.format() for arg in self.arguments]
+
+        # Remove empty optionnal arguments
```

Comment:
I don't understand this part. Do you have an example of optional argument?

Reply:
This is to make `strace.py` adopt a similar behaviour to `strace`. I am testing by looking at the opened files of a plain `ls` command.

This is the output on commit `18dfa4002969dae30671c349ba34ff2cc0a1c36b`:
```
python-ptrace> python strace.py -e open,openat ls
execve(/usr/bin/ls, ['/usr/bin/ls'], [/* 40 vars */]) = 6914
openat(-100, '/etc/ld.so.cache', O_RDONLY|O_CLOEXEC (0o2000000), O_RDONLY) = 3
openat(-100, '/usr/lib/libcap.so.2', O_RDONLY|O_CLOEXEC (0o2000000), O_RDONLY) = 3
openat(-100, '/usr/lib/libc.so.6', O_RDONLY|O_CLOEXEC (0o2000000), O_RDONLY) = 3
openat(-100, '/usr/lib/locale/locale-archive', O_RDONLY|O_CLOEXEC (0o2000000), O_RDONLY) = 3
openat(-100, '.', O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC (0o2204000), O_RDONLY) = 3
build	 cptrace  doc	    gdb.py	 master   ptrace		  README.rst   setup.cfg	 setup.py   SYSCALL_PROTOTYPES.codegen.py  tests
COPYING  dist	  examples  MANIFEST.in  ppc64le  python_ptrace.egg-info  runtests.py  setup_cptrace.py  strace.py  test_doc.py			   tox.ini
*** Process 6914 exited normally ***
```
And this is strace in the same environment:
```
python-ptrace> strace -e open,openat ls
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libcap.so.2", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
openat(AT_FDCWD, ".", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
build	 cptrace  doc	    gdb.py	 master   ptrace		  README.rst   setup.cfg	 setup.py   SYSCALL_PROTOTYPES.codegen.py  tests
COPYING  dist	  examples  MANIFEST.in  ppc64le  python_ptrace.egg-info  runtests.py  setup_cptrace.py  strace.py  test_doc.py			   tox.ini
+++ exited with 0 +++
```

You can see there is the `O_RDONLY` flag twice in the python version, which is an error as the third argument of `open` is the permissions mask (also the fourth of `openat`).
This fourth argument is 0 by default. When formatted, 0 means no values have been ORed together, so the argument has no text: (When removing the snippet)
```
python-ptrace> python strace.py -e open,openat ls
execve(/usr/bin/ls, ['/usr/bin/ls'], [/* 40 vars */]) = 7347
openat(AT_FDCWD, '/etc/ld.so.cache', O_RDONLY|O_CLOEXEC (0o2000000), ) = 3
openat(AT_FDCWD, '/usr/lib/libcap.so.2', O_RDONLY|O_CLOEXEC (0o2000000), ) = 3
openat(AT_FDCWD, '/usr/lib/libc.so.6', O_RDONLY|O_CLOEXEC (0o2000000), ) = 3
openat(AT_FDCWD, '/usr/lib/locale/locale-archive', O_RDONLY|O_CLOEXEC (0o2000000), ) = 3
openat(AT_FDCWD, '.', O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC (0o2204000), ) = 3
build	 cptrace  doc	    gdb.py	 master   ptrace		  README.rst   setup.cfg	 setup.py   SYSCALL_PROTOTYPES.codegen.py  tests
COPYING  dist	  examples  MANIFEST.in  ppc64le  python_ptrace.egg-info  runtests.py  setup_cptrace.py  strace.py  test_doc.py			   tox.ini
*** Process 7347 exited normally ***
```

The code then "strips" the argument list to avoid having empty parameters, as does `strace`.